src/Templates/BlockTemplatesCompatibility.php

```diff
@@ -448,38 +459,35 @@ private static function has_single_product_template_blocks( $parsed_blocks, $sin
 
 
 	/**
-	 * Create a dictionary of blocks split by type:
-	 * template parts before the blocks.
-	 * blocks (they can include a template part).
-	 * template parts after the blocks.
+	 * Group blocks in this way:
+	 * B1 + TP1 + B2 + B3 + B4 + TP2 + B5
+	 * (B = Block, TP = Template Part)
+	 * becomes:
+	 * [[B1], [TP1], [B2, B3, B4], [TP2], [B5]]
 	 *
 	 * @param array $parsed_blocks Array of parsed block objects.
-	 * @return array Array of blocks split by type.
+	 * @return array Array of blocks grouped by template part.
 	 */
-	private static function split_blocks_by_type( $parsed_blocks ) {
+	private static function group_blocks( $parsed_blocks ) {
 		return array_reduce(
 			$parsed_blocks,
 			function( $carry, $block ) {
 				if ( 'core/template-part' === $block['blockName'] ) {
-					if ( empty( $carry['blocks'] ) ) {
-						array_push( $carry['pre-blocks-template-parts'], $block );
-						return $carry;
-					} else {
-						array_push( $carry['post-blocks-template-parts'], $block );
-						return $carry;
-					}
+					array_push( $carry, array( $block ) );
+					return $carry;
 				}
 				if ( empty( $block['blockName'] ) ) {
 					return $carry;
 				}
-				array_push( $carry['blocks'], $block );
+				$last_element_index = count( $carry ) - 1;
+				if ( isset( $carry[ $last_element_index ][0]['blockName'] ) && 'core/template-part' !== $carry[ $last_element_index ][0]['blockName'] ) {
+					array_push( $carry[ $last_element_index ], $block );
+					return $carry;
+				}
```

Comment:
I guess you can use the [`end()` PHP function](https://www.php.net/end) here to simplify the code:

```PHP
$last_element = end( $carry );
if ( isset( $last_element[0]['blockName'] ) && 'core/template-part' !== $last_element[0]['blockName'] ) {
	array_push( $last_element, $block );
	return $carry;
}
```

Reply:
Mmm, I tried with your suggestion but the test failed. I suspect it happens because we don't push the updated element in the carry.

So, I prefer to not change anything here. What do you think?

Reply:
It sounds good, I might have missed something. If you keep it as-is, I think you'll need to add an extra check in the case that `count( $carry )` is 0, because then the code will try to use `$carry[ -1 ]` and produce a warning.